tools/build_script_generator/cmake/resources/CMakeLists.txt.in

```diff
@@ -24,11 +24,9 @@ INCLUDE_DIRECTORIES(
 )
 INCLUDE_DIRECTORIES(SYSTEM include)
 
-%% if "cpp.define" in metadata
-%% for define in metadata["cpp.define"]
+%% for define in flags["CPPDEFINES"]["base"]
 SET_TARGET_PROPERTIES($(CMAKE_PROJECT_NAME).elf PROPERTIES COMPILE_DEFINITIONS "$({{ define }})")
 %% endfor
-%% endif
```

Comment:
@siy How would I add the debug defines to the debug build in CMake? They're in `flags["CPPDEFINES"]["debug"]`.

In the SCons Builder, I just pass the profile via command line: `scons build profile=debug` which is then chosen via Python:

```py
profile = ARGUMENTS.get("profile", "release")
if profile == "debug":
    env.Append(CPPDEFINES = [
        "MODM_DEBUG_BUILD",
    ])
env["CCFLAGS"] = [
    "-fdata-sections",
   # ...
]
if profile == "release":
    env.Append(CCFLAGS = [
        "-Os",
    ])
if profile == "debug":
    env.Append(CCFLAGS = [
        "-Og",
    ])
```

Reply:
CMake uses the `CMAKE_BUILD_TYPE` variable to select different build profiles. It passed to cmake via the `-D` command line option, e.g. `-DCMAKE_BUILD_TYPE=Debug`, similar to the SCons Builder. Depending on CMAKE_BUILD_TYPE different compiler flags will be chosen.

In any case the flags in CMAKE_C_FLAGS/CMAKE_CXX_FLAGS are used. Profile specific flags can be specified in CMAKE_CXX_FLAGS_* (e.g. CMAKE_CXX_FLAGS_DEBUG for Debug). These will be appended to the default flags in CMAKE_CXX_FLAGS.

It is perfectly legal for the user to not specify CMAKE_BUILD_TYPE. In this case the variable is undefined and only the flags in CMAKE_CXX_FLAGS are considered. You'll probably have to set CMAKE_BUILD_TYPE yourself in the CMake file to default to a specific profile (https://cmake.org/pipermail/cmake/2006-February/008371.html).

CMake defines four possible values for CMAKE_BUILD_TYPE: Debug, Release, RelWithDebInfo and MinSizeRel. Others are not possible.

Reply:
Exactly. I think that it will be enough to set CMAKE_C_FLAGS_DEBUG/CMAKE_CXX_FLAGS_DEBUG. Particular profile then can be selected by user or IDE by passing appropriate command line flag to CMake.

Reply:
Thanks, I've adapted the flag settings, defaulted to release build and added debug defines.